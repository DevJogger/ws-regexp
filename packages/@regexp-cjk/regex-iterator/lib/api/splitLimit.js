"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitLimit = void 0;
const _each_1 = require("./_each");
const greedySplit_1 = require("../util/split/greedySplit");
function _pushNonEmpty(arr, s, allowEmpty) {
    if (s !== '' || allowEmpty === true) {
        arr.push(s);
    }
}
function splitLimit(input, separator, limit, options) {
    if (typeof limit === 'object' && limit !== null) {
        options = limit;
        limit = options === null || options === void 0 ? void 0 : options.limit;
    }
    if (limit === 0 || limit < 0 || isNaN(limit)) {
        limit = void 0;
    }
    if (typeof separator === 'string') {
        return (0, greedySplit_1.greedySplit)(input, separator, limit);
    }
    options = options !== null && options !== void 0 ? options : {};
    let ret = [];
    let lastIndex = 0;
    let lastString = '';
    options.global = true;
    let { useFullMatched, allowEmpty, excludeSubMatched, limitMode } = options;
    useFullMatched = !!useFullMatched;
    allowEmpty = !!allowEmpty;
    excludeSubMatched = !!excludeSubMatched;
    let size = 0;
    let row;
    for (row of (0, _each_1._each)(input, separator, options)) {
        let { match, re } = row;
        let s;
        s = input.slice(lastIndex, match.index);
        _pushNonEmpty(ret, lastString = s, allowEmpty);
        //console.log(lastIndex, re.lastIndex, match.index)
        if (checkLimit()) {
            lastIndex = match.index;
            break;
        }
        lastIndex = re.lastIndex;
        if (excludeSubMatched !== true && match.length > 1) {
            if (useFullMatched === true) {
                s = match[0];
                _pushNonEmpty(ret, lastString = s, allowEmpty);
            }
            else {
                for (s of match.slice(1)) {
                    _pushNonEmpty(ret, lastString = s, allowEmpty);
                    if (checkLimit()) {
                        break;
                    }
                }
            }
        }
        size++;
        if (checkLimit()) {
            break;
        }
    }
    if (limit > 0) {
        if (!options.excludeRest) {
            let s = input.slice(lastIndex);
            /*
            console.dir(row)
            console.dir(ret)

            console.dir({
                size,
                length: ret.length,
                limit,
            })
             */
            if (!checkLimit()) {
                _pushNonEmpty(ret, s, allowEmpty);
            }
            else {
                ret[ret.length - 1] += s;
            }
        }
    }
    else {
        _pushNonEmpty(ret, input.slice(lastIndex), allowEmpty);
    }
    function checkLimit() {
        let bool;
        if (limitMode === 1) {
            //console.log(limitMode, size === limit)
            bool = (size === limit);
        }
        else {
            //console.log(limitMode, ret.length === limit)
            bool = ret.length === limit;
        }
        return bool;
    }
    return ret;
}
exports.splitLimit = splitLimit;
exports.default = splitLimit;
//# sourceMappingURL=splitLimit.js.map