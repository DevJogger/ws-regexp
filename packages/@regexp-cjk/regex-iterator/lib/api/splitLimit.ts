import _each from './_each';
import { ICloneRegexpOptionsCustom, ICloneRegexpOptions } from '@regexp-cjk/clone-regexp/lib/types';

function greedySplit(str: string, separator: string | RegExp, limit?: number)
{
	const result = str.split(separator, limit);
	if (result.length === limit)
	{
		let length = 0;
		if (typeof separator === 'string')
		{
			length = result.join(separator).length;
		}
		else
		{
			length = result.reduce((l, x, i) =>
			{
				let separatorLength = 0;
				if (i + 1 < limit)
				{
					separatorLength = str.slice(l).match(separator).shift().length;
				}
				return l + x.length + separatorLength;
			}, 0);
		}
		result[limit - 1] += str.slice(length);
	}
	return result;
}

function _pushNonEmpty(arr: string[], s: string, allowEmpty?: boolean)
{
	if (s !== '' || allowEmpty === true)
	{
		arr.push(s)
	}
}

export function splitLimit(input: string, separator: string | RegExp, limit?: number, options: ICloneRegexpOptionsCustom & {
	useFullMatched?: boolean,
	excludeRest?: boolean,
	allowEmpty?: boolean,
} = {}): string[]
{
	if (typeof separator === 'string')
	{
		return greedySplit(input, separator, limit)
	}

	let ret: string[] = []
	let lastIndex = 0;
	let lastString = '';

	(options as ICloneRegexpOptions).global = true;

	let { useFullMatched, allowEmpty } = options;
	useFullMatched = !!useFullMatched;
	allowEmpty = !!allowEmpty;

	for (let row of _each(input, separator, options))
	{
		let { match, re } = row;
		let s: string;

		s = input.slice(lastIndex, match.index)

		_pushNonEmpty(ret, lastString = s, allowEmpty);

		//console.log(lastIndex, re.lastIndex, match.index)

		lastIndex = re.lastIndex;

		if (match.length > 1)
		{
			lastIndex = re.lastIndex;

			if (useFullMatched === true)
			{
				s = match[0];

				_pushNonEmpty(ret, lastString = s, allowEmpty);
			}
			else
			{
				for (s of match.slice(1))
				{
					_pushNonEmpty(ret, lastString = s, allowEmpty);

					if (ret.length === limit)
					{
						break;
					}
				}
			}

		}

		if (ret.length === limit)
		{
			break;
		}
	}

	if (limit > 0)
	{
		if (!options.excludeRest)
		{
			ret[ret.length-1] += input.slice(lastIndex);
		}
	}
	else
	{
		_pushNonEmpty(ret, input.slice(lastIndex), allowEmpty);
	}

	return ret;
}

export default splitLimit
