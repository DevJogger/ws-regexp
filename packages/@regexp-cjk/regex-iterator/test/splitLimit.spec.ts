import splitLimit from '../lib/api/splitLimit';

test(`splitLimit`, () =>
{

	let actual = splitLimit('kkabbcccccccccccfff', /(\w)\1/, 4);

	console.dir(actual)

	expect(actual.includes('')).toBeFalsy()
	expect(actual.length).toBeLessThanOrEqual(4);

});

test(`actual.join will same as source text`, () =>
{
	let input = 'kkabbcccccccccccfff';
	let limit = 6;

	let actual = splitLimit(input, /(.)/, limit);

	checkLimit(actual, limit);

	expect(actual.join('')).toStrictEqual(input);

	actual = splitLimit(input, /(.{2})/, limit);

	checkLimit(actual, limit);

	expect(actual.join('')).toStrictEqual(input);

	actual = splitLimit(input, /(.{3})/, limit);

	expect(actual.join('')).toStrictEqual(input);

	checkLimit(actual, limit);

	actual = splitLimit(input, /(.{4})/, limit);

	expect(actual.join('')).toStrictEqual(input);

	checkLimit(actual, limit);

});

test(`same as string.split`, () =>
{
	let input = 'kkabbcccccccccccfff';
	let re = /(\w{2})/;

	let actual = splitLimit(input, re);
	let expected = input.split(re).filter(s => s !== '');

	expect(actual.includes('')).toBeFalsy()
	expect(actual).toStrictEqual(expected);
	expect(actual).toMatchSnapshot();

	actual = splitLimit(input, re, void 0, {
		allowEmpty: true,
	});
	expected = input.split(re);

	expect(actual).toStrictEqual(expected);
	expect(actual).toMatchSnapshot();

});

function checkLimit(actual: string[], limit: number)
{
	expect(actual.includes('')).toBeFalsy()
	expect(actual.length).toBeLessThanOrEqual(limit);
	expect(actual).toMatchSnapshot();
}
