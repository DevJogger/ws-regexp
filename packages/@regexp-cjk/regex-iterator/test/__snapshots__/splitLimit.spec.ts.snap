// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actual.join will same as source text 1`] = `
Array [
  "k",
  "k",
  "a",
  "b",
  "b",
  "cccccccccccfff",
]
`;

exports[`actual.join will same as source text 2`] = `
Array [
  "kk",
  "ab",
  "bc",
  "cc",
  "cc",
  "ccccccfff",
]
`;

exports[`actual.join will same as source text 3`] = `
Array [
  "kka",
  "bbc",
  "ccc",
  "ccc",
  "ccc",
  "cfff",
]
`;

exports[`actual.join will same as source text 4`] = `
Array [
  "kkab",
  "bccc",
  "cccc",
  "ccccfff",
]
`;

exports[`same as string.split 1`] = `
Array [
  "kk",
  "ab",
  "bc",
  "cc",
  "cc",
  "cc",
  "cc",
  "cc",
  "ff",
  "f",
]
`;

exports[`same as string.split 2`] = `
Array [
  "",
  "kk",
  "",
  "ab",
  "",
  "bc",
  "",
  "cc",
  "",
  "cc",
  "",
  "cc",
  "",
  "cc",
  "",
  "cc",
  "",
  "ff",
  "f",
]
`;
